code/drasil-code/lib/Language/Drasil/Code/Imperative/Doxygen/Import.hs

```diff
@@ -23,6 +23,11 @@ yes, no :: Doc
 yes = text "YES"
 no = text "NO"
 
+-- | Renderings of comments regarding the default tag values for various Doxygen settings.
+tagDefNO, tagDefYES :: String
+tagDefNO = "# The default value is: NO."
+tagDefYES = "# The default value is: YES."
```

Comment:
I was definitely thinking of splitting this into 3: "# The default value is: %s.", "NO" and "YES", where I'm using C-style printf formatting here (but this should be coded differently).  I would then create things like
```haskell
defNo, defYes :: Doc
defNo = defaultVal no
defYes = defaultVal yes
```
with
```haskell
defaultVal :: String -> Doc
```

Reply:
Considering that yes and no and already declared as Doc in the file, is the following acceptable?
```
defNo, defYes :: Doc
defNo = defaultVal no
defYes = defaultVal yes
 
defaultVal :: Doc -> Doc
defaultVal val = text "# The default value is: " <+> val <+> text "."
```

Regarding QDefinition, I understand that this defines an expression with inputs, a definition, and a domain. I am currently reviewing its implementation. I have found some implementations in the code, including those below. Quite a few other QDefinitions are commented out. Can you please point me to an implementation that is similar to the problem in hand?

```
auxiliaryConstants :: [QDefinition]
auxiliaryConstants = assumptionConstants ++ gBRSpecParamVals

func_reqs_req6_pulledList :: [QDefinition]
func_reqs_req6_pulledList = [nonFL, glaTyFac, dimLL, tolPre, tolStrDisFac, strDisFac, hFromt]
```

Reply:
Not that I'm pretty sure that `text "."` already has a combinator for it.  But otherwise, yes, the first block of code is good.

I'm probably asking for too much wrt `QDefinition` at this point. We'd need a proper generator and back-end printer to make this all work. So let's put that part on the back-burner for now.

Reply:
@JacquesCarette Helpers.hs. is very helpful indeed. I tried to import the relevant helper and got the error "Could not load module ‘Language.Drasil.Printing.Helpers’ it is a hidden module in the package ‘drasil-printers-0.1.10.0’" when compiling. 
The helpers weren't already imported somewhere else since ''Variable not in scope" (dot) if nothing is imported. Could you comment on this, why/how are these helpful helper functions hidden, and how are they supposed to be imported? 

I had the same issue with code/drasil-gool/lib/GOOL/Drasil/LanguageRenderer.hs, and code/drasil-printers/lib/Language/Drasil/Plain/Print.hs, which also contain combinators for `text "."`

Reply:
All packages in Drasil only export a limited number of their inner modules. Usually, they will have a base file that just re-exports a lot of it's functionality. To, generally, get a full list of the exported modules of a package, check out it's `package.yaml` file. The files listed under "exposed-modules" under the "library" section would be the ones that you can "import" from external modules. With both `drasil-printers` and `drasil-gool`, they have a limited number of, manually written, exported modules that you can use from external modules. It's only from that exposed modules that would be able to import from. If no "exposed-modules" list is listed in the `package.yaml` file, then it means that all module files in the `lib/` folder are exported, by default.

Since you found that combinator occurs in multiple areas, it might make sense for you to delete all of them in favour of making a single one in `drasil-utils`'s `Utils.Drasil.Document` module. Afterwards, you can import & re-export that new combinator you would have made via the main (and sole) `Utils.Drasil` module file that's exported by `drasil-utils`. Finally, you should be able to `import Utils.Drasil (yourCombinator)` from this file.

Reply:
Thanks @balacij , perfect answer. And the right process: when we notice duplication in Drasil, we fix it (unless there are strong reasons not to).

Reply:
Thanks for the explanation, @balacij. Very informative. 

Regarding the deletion of duplicate combinators, upon deeper review,  there are similar combinators in `Helpers.hs`, `LanguageRenderer.hs`, and `Print.hs`, but there are slight differences between these, for good reason I'm sure, and many other combinators in these files are not duplicated. I am not sure if combining everything into one file is the best action at this time, maybe @JacquesCarette can comment. It might be worth looking into if a cleanup is needed, but I cannot say be certain of something like this at this time.

The appropriate combinator for what I need to do is `dot = text "."` from `LanguageRenderer.hs` in `drasil-gool`. The yaml package only exposes `GOOL.Drasil`, so I think I will simply expose `GOOL.Drasil.LanguageRenderer` as well.

Reply:
An even better combinator would be a "terminate sentence" (which does exist for `Sentence`) one. Finding a name is non-trivial.

As discussion on #2887 brought up too, we need to refactor the printing, not just expose stuff. `text" ."` has nothing to do with GOOL. It happens to be defined there (too) because GOOL used to be stand-alone.

Reply:
I do not see a terminate sentence combinator in `Utils/Drasil/Sentence.hs`, did you want me to add one there?

Reply:
There's both `+:+.` and `!.`. But I think those work on `Sentence`, not `Doc`. You can indeed re-use those names, as code shouldn't be working on `Doc` and `Sentence` at the same time.

Reply:
@JacquesCarette okay, I will re-use the names for Doc. 

Obviously I wouldn't be defining this in Sentence.hs. Where should I add such a terminate sentence combinator for Doc?

Reply:
@balacij is currently refactoring related matters - I think he's in the best position to make suggestions about that.

Reply:
`drasil-utils`'s `Utils.Drasil.Document` looks like a good location to me. You might also need to re-import your combinator in `Utils.Drasil` and expose it in `Utils.Drasil` to get it working because `Utils.Drasil.Document` is a hidden module.

Depending on #2887, we might move the `Doc`-related combinators into a new package on their own, but that won't affect this much.